plugins/workspace-backpack/src/index.js

```diff
@@ -289,19 +353,123 @@ export class Backpack {
         'translate(' + this.left_ + ',' + this.top_ + ')');
   }
 
+  /**
+   * Returns the count of items in the backpack.
+   * @return {number} The count.
+   */
+  getCount() {
+    return this.contents_.length;
+  }
+
+  /**
+   * Returns backpack contents XML.
+   * @return {!Array<string>} The backpack contents.
+   */
+  getContents() {
+    // Return a shallow copy of the contents array.
+    return [...this.contents_];
+  }
+
+  /**
+   * Empties the backpack's contents. If the contents-flyout is currently open
+   * it will be closed.
+   */
+  empty() {
+    if (!this.getCount()) {
+      return;
+    }
+    this.contents_.length = 0;
+    // TODO: Fire UI event for Backpack content change.
+    this.close();
+  }
+
+  /**
+   * Handles a block drop on this backpack.
+   * @param {!Blockly.BlockSvg} block The block being dropped on the backpack.
+   */
+  handleBlockDrop(block) {
+    const blockXml = Blockly.Xml.blockToDomWithXY(block);
+    this.addItem(cleanBlockXML(blockXml));
+  }
+
+  /**
+   * Adds item to backpack.
+   * @param {string} item Text representing the XML tree of a block to add,
+   * cleaned of all unnecessary attributes.
+   */
+  addItem(item) {
+    if (this.contents_.indexOf(item) !== -1) {
+      return;
+    }
+
+    this.contents_.unshift(item);
+    while (this.contents_.length > this.maxItems_) {
+      this.contents_.pop();
+    }
+    // TODO: Fire UI event for Backpack content change.
+  }
+
+  /**
+   * Sets backpack contents XML.
+   * @param {!Array<string>} contents The new backpack contents.
+   */
+  setContents(contents) {
+    this.contents_.length = 0;
```

Comment:
Both approaches work. The benefit to setting the `length` to 0 is that if there was any references to `this.contents_`, this set would affect all references to the array, while setting this variable to an empty array would be reassigning the reference to a new array.

Not related to that question, I this this clear of the array is actually unnecessary since I'm setting `this.contents_` to a new array right afterwards.